docs/SPEC.md

```diff
@@ -788,6 +795,23 @@ Array literals construct a value with the array type.
     ArrayLiteral   = "[" ExpressionList "]" .
     ExpressionList = [ Expression { "," Expression } ] .
 
+##### Dictionary literals
+
+Dictionary literals construct a value with the dict type.
+
+    DictLiteral     = "|" AssociativeList "|" .
+    AssociativeList = [ Association { "," AssociativeList } ] .
+    Association     = Expression ":" Expression .
+
+The keys can be arbitrary expressions.
+The type system will enforce that all keys are of the same type.
+
+Examples:
+
+    a = "a"
+    c = | a: 1, "b": 2 | // dictionary mapping string values to integers
+    d = | a: 1, 2: 3 | // type error: cannot mix string and integer keys
```

Comment:
I like this new syntax, its minimal and clear enough that we have some kind of data type being defined.